libs/core/src/mutex.cpp

```diff
@@ -0,0 +1,94 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/mutex.hpp"
+
+namespace fetch {
+namespace {
+Mutexregister mutex_register;
+}
+
+void RegisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread)
+{
+  mutex_register.RegisterMutexAcquisition(mutex, thread);
+}
+
+void UnregisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread)
+{
+  mutex_register.UnregisterMutexAcquisition(mutex, thread);
+}
+
+void FindDeadlock(DebugMutex *mutex, std::thread::id thread)
+{
+  mutex_register.FindDeadlock(mutex, thread);
+}
+
+void Mutexregister::RegisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread)
+{
+  std::lock_guard<std::mutex> guard(mutex_);
```

Comment:
`FETCH_LOCK`?

Reply:
`FETCH_LOCK` cannot be used here as that would lead to infinite recursion.